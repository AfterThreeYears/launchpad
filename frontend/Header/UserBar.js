/* @flow */

import React, { Component } from 'react';
import type { User } from '../types';
import HeaderDropdownButton from './HeaderDropdownButton';
import HeaderDropdown from './HeaderDropdown';
import HeaderButton from './HeaderButton';
import './UserBar.less';

export default class UserBar extends Component {
  props: {|
    user: ?User,
    onLogin: () => any,
    onLogout: () => any,
  |};

  state: {|
    userDropdownOpen: boolean,
  |};

  constructor() {
    super();
    this.state = {
      userDropdownOpen: false,
    };
  }

  handleLogin = (e: Event) => {
    e.preventDefault();
    this.props.onLogin();
  };

  handleLogout = (e: Event) => {
    e.preventDefault();
    this.props.onLogout();
  };

  handleOpenUserDropdown = (e: Event) => {
    e && e.preventDefault();

    this.setState({
      userDropdownOpen: true,
    });
  };

  handleCloseUserDropdown = (e?: Event) => {
    e && e.preventDefault();

    this.setState({
      userDropdownOpen: false,
    });
  };

  renderButton(): React$Element<*> {
    let onClick;
    if (this.state.userDropdownOpen) {
      onClick = this.handleCloseUserDropdown;
    } else {
      onClick = this.handleOpenUserDropdown;
    }
    return (
      <HeaderButton active={this.state.userDropdownOpen} onClick={onClick}>
        <span className="icon-user" />
        <span>
          {(this.props.user && this.props.user.githubUsername) || 'no-username'}
        </span>
        <span className="UserBar-Arrow">â–¾</span>
      </HeaderButton>
    );
  }

  render() {
    if (this.props.user) {
      return (
        <div className="UserBar">
          <HeaderDropdown
            open={this.state.userDropdownOpen}
            onClose={this.handleCloseUserDropdown}
            button={this.renderButton()}
          >
            <HeaderDropdownButton onClick={this.handleLogout}>
              Sign out
            </HeaderDropdownButton>
            <HeaderDropdownButton>
              <a href="/list">Pads you created</a>
            </HeaderDropdownButton>
          </HeaderDropdown>
        </div>
      );
    } else {
      return (
        <div className="UserBar">
          <HeaderButton onClick={this.handleLogin}>
            <span className="icon-github" />
            <span>Sign in</span>
          </HeaderButton>
        </div>
      );
    }
  }
}
