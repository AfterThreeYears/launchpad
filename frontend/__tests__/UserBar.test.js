/* @flow */

/**
 * @jest-environment node
 */

import React from 'react';
import { StaticRouter, HashRouter } from 'react-router-dom';
import UserBar from '../Header/UserBar';
import { mount } from 'enzyme';
import serializer from 'enzyme-to-json/serializer';

expect.addSnapshotSerializer(serializer);

test('renders normal', () => {
  const component = mount(
    <HashRouter>
      <div>
        <UserBar
          user={{
            id: 'test-id',
            githubUsername: 'example',
            pads: [],
          }}
          onLogin={() => {}}
          onLogout={() => {}}
        />,
      </div>
    </HashRouter>,
  );
  expect(component).toMatchSnapshot();

  component.find('HeaderButton').simulate('click');
  expect(component).toMatchSnapshot();

  component.find('HeaderButton').simulate('click');
  expect(component).toMatchSnapshot();
});