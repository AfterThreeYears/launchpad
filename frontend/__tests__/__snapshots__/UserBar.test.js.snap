// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders normal 1`] = `
<HashRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 2,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <div>
      <UserBar
        onLogin={[Function]}
        onLogout={[Function]}
        user={
          Object {
            "githubUsername": "example",
            "id": "test-id",
            "pads": Array [],
          }
        }
      >
        <div
          className="UserBar"
        >
          <HeaderDropdown
            button={
              <HeaderButton
                active={false}
                onClick={[Function]}
              >
                <span
                  className="icon-user"
                />
                <span>
                  example
                </span>
                <span
                  className="UserBar-Arrow"
                >
                  ▾
                </span>
              </HeaderButton>
            }
            onClose={[Function]}
            open={false}
          >
            <div
              className="HeaderDropdown"
            >
              <HeaderButton
                active={false}
                onClick={[Function]}
              >
                <button
                  className="HeaderButton"
                  onClick={[Function]}
                >
                  <div
                    className="HeaderButton-Inner"
                  >
                    <span
                      className="icon-user"
                    />
                    <span>
                      example
                    </span>
                    <span
                      className="UserBar-Arrow"
                    >
                      ▾
                    </span>
                  </div>
                </button>
              </HeaderButton>
              <div
                className="HeaderDropdown-Dropdown"
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <HeaderDropdownButton
                  onClick={[Function]}
                >
                  <button
                    className="HeaderDropdownButton"
                    onClick={[Function]}
                  >
                    Sign out
                  </button>
                </HeaderDropdownButton>
                <HeaderDropdownButton>
                  <button
                    className="HeaderDropdownButton"
                  >
                    <Link
                      replace={false}
                      to="/list"
                    >
                      <a
                        href="#/list"
                        onClick={[Function]}
                      >
                        Pads you created
                      </a>
                    </Link>
                  </button>
                </HeaderDropdownButton>
              </div>
            </div>
          </HeaderDropdown>
        </div>
      </UserBar>
      ,
    </div>
  </Router>
</HashRouter>
`;

exports[`renders normal 2`] = `
<HashRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 2,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <div>
      <UserBar
        onLogin={[Function]}
        onLogout={[Function]}
        user={
          Object {
            "githubUsername": "example",
            "id": "test-id",
            "pads": Array [],
          }
        }
      >
        <div
          className="UserBar"
        >
          <HeaderDropdown
            button={
              <HeaderButton
                active={true}
                onClick={[Function]}
              >
                <span
                  className="icon-user"
                />
                <span>
                  example
                </span>
                <span
                  className="UserBar-Arrow"
                >
                  ▾
                </span>
              </HeaderButton>
            }
            onClose={[Function]}
            open={true}
          >
            <div
              className="HeaderDropdown"
            >
              <HeaderButton
                active={true}
                onClick={[Function]}
              >
                <button
                  className="HeaderButton HeaderButton--active"
                  onClick={[Function]}
                >
                  <div
                    className="HeaderButton-Inner"
                  >
                    <span
                      className="icon-user"
                    />
                    <span>
                      example
                    </span>
                    <span
                      className="UserBar-Arrow"
                    >
                      ▾
                    </span>
                  </div>
                </button>
              </HeaderButton>
              <div
                className="HeaderDropdown-Dropdown"
                style={Object {}}
              >
                <HeaderDropdownButton
                  onClick={[Function]}
                >
                  <button
                    className="HeaderDropdownButton"
                    onClick={[Function]}
                  >
                    Sign out
                  </button>
                </HeaderDropdownButton>
                <HeaderDropdownButton>
                  <button
                    className="HeaderDropdownButton"
                  >
                    <Link
                      replace={false}
                      to="/list"
                    >
                      <a
                        href="#/list"
                        onClick={[Function]}
                      >
                        Pads you created
                      </a>
                    </Link>
                  </button>
                </HeaderDropdownButton>
              </div>
            </div>
          </HeaderDropdown>
        </div>
      </UserBar>
      ,
    </div>
  </Router>
</HashRouter>
`;

exports[`renders normal 3`] = `
<HashRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 2,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <div>
      <UserBar
        onLogin={[Function]}
        onLogout={[Function]}
        user={
          Object {
            "githubUsername": "example",
            "id": "test-id",
            "pads": Array [],
          }
        }
      >
        <div
          className="UserBar"
        >
          <HeaderDropdown
            button={
              <HeaderButton
                active={false}
                onClick={[Function]}
              >
                <span
                  className="icon-user"
                />
                <span>
                  example
                </span>
                <span
                  className="UserBar-Arrow"
                >
                  ▾
                </span>
              </HeaderButton>
            }
            onClose={[Function]}
            open={false}
          >
            <div
              className="HeaderDropdown"
            >
              <HeaderButton
                active={false}
                onClick={[Function]}
              >
                <button
                  className="HeaderButton"
                  onClick={[Function]}
                >
                  <div
                    className="HeaderButton-Inner"
                  >
                    <span
                      className="icon-user"
                    />
                    <span>
                      example
                    </span>
                    <span
                      className="UserBar-Arrow"
                    >
                      ▾
                    </span>
                  </div>
                </button>
              </HeaderButton>
              <div
                className="HeaderDropdown-Dropdown"
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <HeaderDropdownButton
                  onClick={[Function]}
                >
                  <button
                    className="HeaderDropdownButton"
                    onClick={[Function]}
                  >
                    Sign out
                  </button>
                </HeaderDropdownButton>
                <HeaderDropdownButton>
                  <button
                    className="HeaderDropdownButton"
                  >
                    <Link
                      replace={false}
                      to="/list"
                    >
                      <a
                        href="#/list"
                        onClick={[Function]}
                      >
                        Pads you created
                      </a>
                    </Link>
                  </button>
                </HeaderDropdownButton>
              </div>
            </div>
          </HeaderDropdown>
        </div>
      </UserBar>
      ,
    </div>
  </Router>
</HashRouter>
`;
